src/main/java/com/mixpanel/android/viewcrawler/ViewVisitor.java

```diff
@@ -152,19 +180,111 @@ public void accumulate(View found) {
                 LayoutRule originalValue = new LayoutRule(newVerb, currentRules[newVerb]);
                 mOriginalValues.put(found, originalValue);
             }
-            setLayout(found, newVerb, newAnchorId);
+
+            try {
+                setLayout(found, newVerb, newAnchorId);
+            } catch (LayoutUpdateException e) {
+                throw e;
+            }
         }
 
         private void setLayout(View target, int verb, int anchorId) {
             RelativeLayout.LayoutParams params = (RelativeLayout.LayoutParams)target.getLayoutParams();
             params.addRule(verb, anchorId);
+
+            final ArrayList<Integer> rules;
+            if (mHorizontalRules.contains(verb)) {
+                rules = mHorizontalRules;
+            } else if (mVerticalRules.contains(verb)) {
+                rules = mVerticalRules;
+            } else {
+                rules = null;
+            }
+
+            if (rules != null && !verifyLayout(target, rules)) {
+                JSONObject errorInfo = new JSONObject();
+                try {
+                    errorInfo.put("cid", mName);
+                    errorInfo.put("error_type", "circular_dependency");
+                } catch (JSONException e) {
+                    ; // won't reach here
+                }
+                throw new LayoutUpdateException("Circular dependency detected!", errorInfo);
+            }
+
             target.setLayoutParams(params);
         }
 
+        private boolean verifyLayout(View target, ArrayList<Integer> rules) {
+            ViewGroup parent = (ViewGroup) target.getParent();
+            SparseArray<View> idToChild = new SparseArray<View>();
+
+            for (int i = 0; i < parent.getChildCount(); i++) {
+                View child = parent.getChildAt(i);
+                idToChild.put(child.getId(), child);
+            }
+
+            ArrayMap<View, ArrayList<View>> dependencyGraph = new ArrayMap<View, ArrayList<View>>();
```

Comment:
I didn't really know a good structure for this, so that I was following the one used in Android source code. I think the point is that ArrayMap is more memory efficient and since there won't be too many nodes in the graph, it's almost as fast as regular HashMap.